2-WebApp-graph-user/2-1-Call-MSGraph/Startup.cs

```diff
@@ -41,6 +40,16 @@ public void ConfigureServices(IServiceCollection services)
                .AddMsal(Configuration, new string[] { Constants.ScopeUserRead })
                .AddInMemoryTokenCaches();
 
+/*
+            // The following code enables client certificate (instead of client secret)
+            services.Configure<ConfidentialClientApplicationOptionsWithClientCertificate>(options =>
+                {
+                        options.ClientSecret = null;
+                options.ClientCertificate = LoadCertificate();
```

Comment:
Having an ICertificateLoader (or ICertificateManager or something like that) interface here would be better than IOptions since it's not serializable.  Update "AddMsal()" to add a parameter with Type which can default to null.  If it's null, we'll provide a default implementation that doesn't load a certificate (the "i'm not using a certificate" case).  If it's not null, then we'll register the Type (assuming it derives from ICertificateLoader) with the DI container so they can load the certificate.  The only question I have is do we need more than one certificate for a service (maybe) in which case we may want the LoadCertificate() method to take a string parameter with a name...?

Reply:
Something along the same lines from me too. I'd strongly suggest we keep ourselves out of the business of loading certificates by name or thumbprint or similar approaches. The develoepr might have that cert in the personal or machine store, or will pull from a file in disk or keyvault. I am more thinking towards an overload like following:

AddMsal(this IServiceCollection services, IConfiguration configuration, IEnumerable<string> initialScopes, X509Certificate2 certificate, string configSectionName = "AzureAd")

Also, the overload will throw an exception if both secret and cert are defined.
 

---
In reply to: [326277037](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/189#discussion_r326277037) [](ancestors = 326277037)